code/scripting/hook_conditions.cpp

```diff
@@ -0,0 +1,274 @@
+
+	template<typename _conditions_t, typename _operating_t, typename _cache_t> friend class EvaluatableConditionImpl;
+public:
+	std::unique_ptr<EvaluatableCondition> parse(const SCP_string& input) const override {
+		return ::make_unique<EvaluatableConditionImpl<conditions_t, operating_t, cache_t>>(*this, input);
+	}
+
+	ParseableConditionImpl(SCP_string documentation_, const operating_t conditions_t::* object_, std::function<cache_t(const SCP_string&)> cache_, std::function<bool(operating_t, const cache_t&)> evaluate_) :
+		ParseableCondition(std::move(documentation_)), object(object_), cache(std::move(cache_)), evaluate(std::move(evaluate_)) { }
+};
+
+template<typename conditions_t, typename operating_t, typename cache_t>
+class EvaluatableConditionImpl : public EvaluatableCondition {
+	const ParseableConditionImpl<conditions_t, operating_t, cache_t>& condition;
+	cache_t cached;
+
+public:
+	EvaluatableConditionImpl(const ParseableConditionImpl<conditions_t, operating_t, cache_t>& _condition, const SCP_string& input) : condition(_condition), cached(condition.cache(input)) { }
+
+	bool evaluate(const linb::any& conditionContext) const override {
+		const conditions_t& conditions = linb::any_cast<conditions_t>(conditionContext);
+		return condition.evaluate(conditions.*(condition.object), cached);
+	}
+};
+
+
+namespace hooks {
+
+// ---- Hook Condition System Utility and Parsing methods ----
+
+static bool conditionCompareShip(const ship* shipp, const SCP_string& value) {
+	return shipp != nullptr && stricmp(shipp->ship_name, value.c_str()) == 0;
+}
+
+static bool conditionCompareShipType(const ship* shipp, const int& value) {
+	return shipp != nullptr && Ship_info[shipp->ship_info_index].class_type == value;
+}
+
+static bool conditionCompareShipClass(const ship* shipp, const int& value) {
+	return shipp != nullptr && shipp->ship_info_index == value;
+}
+
+static bool conditionCompareWeaponClass(const weapon* wep, const int& value) {
+	return wep != nullptr && wep->weapon_info_index == value;
+}
+
+static bool conditionIsObjecttype(const object* objp, const int& value) {
+	return objp != nullptr && objp->type == value;
+}
+
+template<typename fnc_t, typename value_t>
+static bool conditionObjectIsShipDo(fnc_t fnc, const object* objp, const value_t& value) {
+	if (objp != nullptr && objp->type == OBJ_SHIP) {
+		return fnc(&Ships[objp->instance], value);
+	}
+	return false;
+}
+
+template<typename fnc_t, typename value_t>
+static bool conditionObjectIsWeaponDo(fnc_t fnc, const object* objp, const value_t& value) {
+	if (objp != nullptr && objp->type == OBJ_WEAPON) {
+		return fnc(&Weapons[objp->instance], value);
+	}
+	return false;
+}
+
+
+static SCP_string conditionParseString(const SCP_string& name) {
+	return name;
+}
+
+static int conditionParseActionId(const SCP_string& name) {
+	for (int i = 0; i < (int)Control_config.size(); i++) {
+		if (SCP_string_lcase_equal_to{}(Control_config[i].text, name))
```

Comment:
Why is this `SCP_string_lcase_equal_to{}` rather than `SCP_string_lcase_equal_to()`?

Reply:
That's just a matter of preference really.
You need an object for the operator() since it's not static, and I prefer to initialize trivial types without constructors with brace-enclosed initializers.
The line could just as well be ``if (SCP_string_lcase_equal_to()(Control_config[i].text, name))`` with no functional difference

Reply:
I didn't think `operator()` would let you do it the other way.  Gotcha.